src/prototype_creep_startup_tasks.js

```diff
@@ -336,6 +336,39 @@ Creep.prototype.repairStructure = function() {
   return false;
 };
 
+/**
+ *
+ * @param {Resource} target Resource object to pick up
+ * @return {number} total received resources amount
+ */
+Creep.prototype.pickupOrWithdrawFromSourcer = function(target) {
+  const creepFreeSpace = this.carryCapacity - _.sum(this.carry);
+  let pickedUp = 0;
+  // this.log('pickupOrWithdrawFromSourcer free '+creepFreeSpace+' '+target+' '+target.amount)
+  if (target.amount < creepFreeSpace) {
+    let container = target.pos.lookFor(LOOK_STRUCTURES).find(function(structure) {
+      return structure.structureType === STRUCTURE_CONTAINER && structure.store[target.resourceType] > 0 && _.sum(structure.store) === structure.storeCapacity;
+    });
+    if (container) {
+      const toWithdraw = Math.min(creepFreeSpace - target.amount, container.store[target.resourceType]);
+      this.withdraw(container, target.resourceType, toWithdraw);
+      pickedUp += toWithdraw;
+    } else {
+      let sourcer = target.pos.lookFor(LOOK_CREEPS).find(function(creep) {
+        return creep.memory && creep.memory.role === 'sourcer' && creep.carry[target.resourceType] > 0 && _.sum(creep.carry) === creep.carryCapacity;
+      });
+      if (sourcer) {
+        const toWithdraw = Math.min(creepFreeSpace - target.amount, sourcer.carry[target.resourceType]);
+        sourcer.transfer(this, target.resourceType, toWithdraw);
+        pickedUp += toWithdraw;
+      }
+    }
+  }
+  this.pickup(target);
```

Comment:
Acording to http://docs.screeps.com/simultaneous-actions.html `pickup` isn't involved into any of pipelins, so it should be done in one tick with `withdraw`.
Nevertheless, if we do `pickup` in first tick and `withdraw` in second, it woun't work, as on the second tick we will already have new 10 energy for picking up. 1 energy unit isn't worth much code complication